versioned_hdf5/hashtable.py

```diff
@@ -124,22 +124,23 @@ def hash(self, data: np.ndarray):
         # Object dtype arrays store the ids of the elements, which may or may not be
         # reused, making it unsuitable for hashing. Instead, we need to make a combined
         # hash with the value of each element.
-        if data.dtype == "object":
-            hash_value = self.hash_function()
+        hash_value = self.hash_function()
+        if data.dtype == object:
             for value in data.flat:
                 if isinstance(value, str):
-                    # default to utf-8 encoding since it's a superset of ascii (the only other
-                    # valid encoding supported in h5py)
+                    # default to utf-8 encoding since it's a superset of ascii (the only
+                    # other valid encoding supported in h5py)
                     value = value.encode("utf-8")
-                # hash the length of value ('n' is ssize_t, which matches the internal type for lengths)
-                hash_value.update(struct.pack("n", len(value)))
+                # hash the length of value ('Q' is unsigned long long, which is 64 bit
+                # on all of today's architectures)
+                hash_value.update(struct.pack("Q", len(value)))
+                # Hash the buffer of bytes, bytearray, memoryview, etc.
```

Comment:
This can trigger very nasty hash collisions if someone has the weird idea to create an array of numpy arrays. Hopefully, h5py should reject it.